File: mgsAnalysis/tms_efield/simulate_simnibs.m
clear; close all; clc;
addpath("/users/mrugank/Applications/SimNIBS-4.0/matlab_tools/")

sub = 1;
subjID = num2str(sub, '%02d');
m2mfoldpath = ['/d/DATC/datc/MD_TMS_EEG/SIMNIBS_output/sub' ...
                subjID '/m2m_sub' subjID];
simfoldpath = ['/d/DATC/datc/MD_TMS_EEG/SIMNIBS_output/sub' ...
                subjID '/simnibs_simulation'];
S = sim_struct('SESSION');
S.fnamehead = [subjID '.msh'];
S.subpath = m2mfoldpath;
S.pathfem = simfoldpath;
S.open_in_gmsh=true;
S.map_to_vol=true;
S.tissues_in_niftis = 'all';
S.fields= 'eEjJvDs';
S.poslist{1} = sim_struct('TMSLIST');
S.poslist{1}.fnamecoil = 'MagVenture_Cool-B70.ccd';
S.poslist{1}.pos(1).centre = 'C3';
S.poslist{1}.pos(1).pos_ydir = 'CP3';
S.poslist{1}.pos(1).distance = 4;

File: mgsAnalysis/tms_efield/simulate_simnibs.py
from simnibs import sim_struct, brainsight, run_simnibs
import os
#subs = [1, 3, 5, 6, 7, 8, 11, 12, 13, 14, 15, 16, 17, 18, 22, 23, 24, 25, 26, 27]
subs = [1, 3, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 22, 23, 24, 25, 26, 27]

# stim_intensities = [55, 47, 56, 63, 60, 53, 54, 55, 60, 44, 46, 58, 47, 55, 50, 53, 50, 48, 50, 50]
# stim_dIdt = [round((a / 60) * 87 *1e6) for a in stim_intensities]

stim_dIdt = [1e6 for a in subs]

coilpath = '/Users/mrugank/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/resources/coil_models/Drakaki_BrainStim_2022/MagVenture_Cool-B70.ccd'
masterpath = "/d/DATD/datd/MD_TMS_EEG/SIMNIBS_output/"
brainsightpath = masterpath+"session_txts/"

#s.add_tmslist(tms_list_targets)
for idx in range(len(subs)):
    sub = subs[idx]
    sub_id = f"sub{sub:02d}"
    navfile = f"{brainsightpath}{sub_id}.txt"
    tms_list_targets, tms_list_samples = brainsight().read(navfile, stim_dIdt[idx])

    # print(tms_list_samples)
    m2mfoldpath = f"{masterpath}{sub_id}/m2m_{sub_id}"
    #simfoldpath = f"{masterpath}{sub_id}/simulation"
    simfoldpath = f"{masterpath}{sub_id}/simstandard"

    if not os.path.exists(simfoldpath):
        os.mkdir(simfoldpath)

        S = sim_struct.SESSION()
        S.fnamehead = os.path.join(m2mfoldpath, f"{sub_id}.msh")
        S.subpath = m2mfoldpath
        S.pathfem = simfoldpath
        S.open_in_gmsh = False
        S.map_to_surf = True
        S.map_to_vol = True
        S.map_to_fsavg = False
        S.map_to_mni = False
        S.fields= 'eEjJvDs'
        #S.tissues_in_niftis = "all"
        tmslist = S.add_tmslist(tms_list_targets)
        tmslist.fnamecoil = coilpath 
        run_simnibs(S, cpus = 4)